### PR TITLE
feat: await primary module change

### PR DESCRIPTION
**Pull request by m1sterc001guy**
This PR adds a generic way of awaiting a primary modules output to be finalized. Often times, after the output outcome happens on the server side, the client needs to wait on some additional operations on the client side (like claiming ecash notes as change from a Lightning transaction).

Right now we don't have a good way to wait on change. This PR adds a function `await_primary_module_output_finalized` that awaits on an outpoint to be finalized. This PR also changes `finalize_and_submit_transaction` to return an `Option<OutPoint>` which will contain the `OutPoint` of the change (if there is change).

Fixes: https://github.com/fedimint/fedimint/issues/2495

### CODE REVIEW

File: fedimint-cli/src/ng.rs [hunk 1]
```diff
@@ -128,7 +128,12 @@ pub async fn handle_ng_command(
             while let Some(update) = updates.next().await {
                 match update {
                     LnReceiveState::Claimed { txid } => {
-                        client.await_claim_notes(operation_id, txid).await?;
+                        client
+                            .await_primary_module_output_finalized(
```
Comment (m1sterc001guy) [inline fedimint-cli/src/ng.rs:132]: This should probably be moved to `subscribe_to_ln_receive_updates`

File: fedimint-client/src/lib.rs [hunk 2]
```diff
@@ -166,15 +167,15 @@ pub trait IGlobalClientContext: Debug + MaybeSend + MaybeSync + 'static {
...
-    ) -> TransactionId;
+    ) -> (TransactionId, Option<OutPoint>);
 
     /// This function is mostly meant for internal use, you are probably looking
     /// for [`DynGlobalClientContext::fund_output`].
     async fn fund_output_dyn(
         &self,
         dbtx: &mut ClientSMDatabaseTransaction<'_, '_>,
         output: InstancelessDynClientOutput,
-    ) -> anyhow::Result<TransactionId>;
```
Comment (elsirion) [inline fedimint-client/src/lib.rs:177]: Nit: we should document what the `OutPoint` represents

File: fedimint-client/src/module/mod.rs [hunk 3]
```diff
@@ -287,6 +288,12 @@ pub trait PrimaryClientModule: ClientModule {
         operation_id: OperationId,
         amount: Amount,
     ) -> ClientOutput<<Self::Common as ModuleCommon>::Output, Self::States>;
+
+    async fn await_primary_module_output_finalized(
+        &self,
+        operation_id: OperationId,
+        out_point: OutPoint,
+    ) -> anyhow::Result<()>;
```
Comment (elsirion) [inline fedimint-client/src/module/mod.rs:296]: I wonder if this should return the amount received? Is that useful or a foot-gun?

**Review by elsirion — APPROVED** [review]
API looks good, didn't look at the state machine changes too closely, can do later.

File: fedimint-client/src/module/mod.rs (hunk 3 above)
Comment (m1sterc001guy) [reply fedimint-client/src/module/mod.rs:296]: I think that would be fine

File: modules/fedimint-ln-client/src/lib.rs [hunk 4]
```diff
@@ -319,42 +336,50 @@ impl LightningClientExt for Client {
...
+            yield LnPayState::Funded;
+
+            match payment_success.await {
+                Ok(preimage) => {
+                    if let Some(change) = change_outpoint {
+                        yield LnPayState::AwaitingChange;
+                        match self.await_primary_module_output_finalized(operation_id, change).await {
+                            Ok(_) => {}
+                            Err(_) => {
+                                yield LnPayState::Failed;
```
Comment (elsirion) [inline modules/fedimint-ln-client/src/lib.rs:356]: There is a nuance here that a malicious federation could block the issuance of e-cash while the contract is funded and the payment succeeds, but probably not a priority to fix.

File: modules/fedimint-ln-client/src/lib.rs [hunk 5]
```diff
@@ -319,42 +336,50 @@ impl LightningClientExt for Client {
...
                 }
+                Err(LightningPayError::Refundable(block_height, error)) => {
+                    yield LnPayState::WaitingForRefund{ block_height, gateway_error: error.clone() };
+
+                    if let Ok(refund_txid) = refund_success.await {
+                        // need to await primary module to get refund
+                        if self.await_primary_module_output_finalized(operation_id, OutPoint{ txid: refund_txid, out_idx: 0}).await.is_ok() {
+                            yield LnPayState::Refunded { gateway_error: error };
+                            return;
+                        }
```
Comment (elsirion) [inline modules/fedimint-ln-client/src/lib.rs:373]: I wonder if we could get rid of the `LightningPayStates::Refunded` state now and instead only rely on `await_primary_module_output_finalized` to determine the high-level operation state (for the output to be finalized the tx has to be accepted). Could be a follow-up.

**Review by elsirion — APPROVED** [review]
Great work! Left some ideas for folow-ups.

**Pull request merged**